File: workers/hub/tasks.py
"""Home & IoT worker — Workstream F.

The Home Assistant adapter and the MQTT publish-back, as Arq jobs.

## What this file is responsible for

The sprint's exit criterion is **indoor and outdoor readings stored every 5–15
minutes**. Workstream E landed the outdoor half; :func:`poll_home_assistant` is
the indoor half, and under ADR 0009 it is the *only* route to indoor
conditions — there is no direct Nest integration and none is planned. If this
job stops, nothing else fills in behind it.

That makes one failure worse than being wrong: **being silent**. A job that
stops running leaves the last readings in place, the Almanac draws a line
through the gap, and a household sees a chart that looks like a steady 21 °C
room. So every run here leaves a mark whether it worked or not —
``integration.last_ok_at`` and ``integration.last_error`` for the current
state, ``job_run`` for the history — and every entity that was asked for and
did not produce a reading is *named* in the job's return value, with a reason.
Nothing is dropped quietly.

## Shape

Every job returns a plain dict rather than writing and staying silent, so a
run's outcome is visible in Arq's result store and in a test with no database
— the same rule Workstream E's jobs follow, and for the same reason: a job
that can only be observed by querying Postgres is a job nobody checks.

Where ``ctx`` carries a ``connection``, rows are persisted through
:mod:`workers.hub.store`; where it does not — mock mode, and every test — the
job still does the work and reports it.

## Cadence

``poll_home_assistant`` is scheduled every five minutes, the fast end of the
sprint's 5–15 minute band, and each ``sensor_source`` then decides for itself
whether it is due via its own ``poll_seconds``. A thermostat is not a soil
probe: the cron sets the *finest* cadence available, the row sets the one it
wants.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from datetime import UTC, datetime
from typing import Any, ClassVar

from workers.runtime import ArqBootstrap

from .entities import SOIL_METRICS, SUPPORTED_METRICS  # noqa: F401  (ADR 0010)
from .health import IntegrationHealth, SourceHealth, health_report, redact
from .mapping import SensorSource, pollable, validation_errors
from .mqtt import (  # noqa: F401  — the topic layout is contract, re-exported
    DISCOVERY_PREFIX,
    MQTT_PREFIX,
    MemoryPublisher,
    availability,
    discovery_messages,
    discovery_topic,
    frost_messages,
    object_id_for,
    publish_all,
    rounds_messages,
    specimen_messages,
    state_topic,
)
from .notify.jobs import (
    notify_frost,
    notify_integration_health,
    notify_rounds,
)
from .settings import HubSettings, get_settings
from .sources.base import HOME_ASSISTANT, HubUnavailable, PollResult

#: The ``integration`` row this worker keeps. ``kind`` is from the schema's
#: CHECK constraint; ``name`` is what the Ministry Office prints.
HA_INTEGRATION = (HOME_ASSISTANT, "Home Assistant")
MQTT_INTEGRATION = ("mqtt", "MQTT broker")


def _settings(ctx: dict[str, Any]) -> HubSettings:
    settings = ctx.get("settings")
    return settings if isinstance(settings, HubSettings) else get_settings()


def _secrets(settings: HubSettings) -> tuple[str, ...]:
    """The values :func:`~workers.hub.health.redact` must remove literally."""
    return tuple(
        value
        for value in (
            settings.ha_token.get_secret_value(),
            settings.mqtt_password.get_secret_value(),
        )
        if value
    )


async def _sources_for(ctx: dict[str, Any]) -> list[SensorSource]:
    """The configured sources: from ``ctx``, then the database, then the mock.

    The ``ctx`` override exists for the API, which may already hold them, and
    for tests. Everything else is the ordinary path.
    """
    provided = ctx.get("sources")
    if provided is not None:
        return [
            item if isinstance(item, SensorSource) else SensorSource.from_row(item)
            for item in provided
        ]
    connection = ctx.get("connection")
    if connection is not None:
        from . import store

        rows = await store.read_sensor_sources(connection)
        return [SensorSource.from_row(dict(row)) for row in rows]
    from . import world

    return world.sensor_sources(_settings(ctx))


async def poll_home_assistant(
    ctx: dict[str, Any], source_id: str | None = None
) -> dict[str, Any]:
    """Read every due sensor source and store what came back.

    ``source_id`` polls one row regardless of whether it is due — what the
    "test this sensor" button in the Ministry Office needs, and what someone
    debugging a mapping wants.

    The return value is the whole story of the run: what was polled, what was
    stored, and every entity that was asked for and did not answer, with a
    reason each. That list is the point of this job as much as the readings
    are.
    """
    from . import store
    from .factory import get_source

    settings = _settings(ctx)
    started_at = datetime.now(UTC)
    secrets = _secrets(settings)

    sources = await _sources_for(ctx)
    problems: list[str] = []
    for source in sources:
        problems.extend(validation_errors(source))

    candidates = pollable(sources)
    if source_id:
        due = [source for source in candidates if source.id == source_id]
        if not due:
            return _report(
                started_at,
                ok=False,
                error=f"no enabled Home Assistant source with id {source_id}",
                sources=len(sources),
            )
    else:
        due = [source for source in candidates if source.is_due(started_at)]

    if not due:
        # Not an error and not a success: nothing was asked of Home Assistant,
        # so nothing was learned about it. Stamping `last_ok_at` here would
        # make a deployment with every source disabled look healthy forever.
        return _report(
            started_at,
            ok=True,
            sources=len(sources),
            polled=0,
            skipped_sources=[source.name for source in candidates],
            problems=problems,
            note="no source was due",
        )

    source_api = ctx.get("source") or get_source()
    result: PollResult = await source_api.poll(due, now=started_at)

    connection = ctx.get("connection")
    stored = 0
    if connection is not None and result.readings:
        stored = await store.write_readings(connection, result.readings)
        touched = sorted({reading.time.date() for reading in result.readings})
        await store.refresh_aggregates(connection, touched[0], touched[-1])

    error = _error_for(result, due, problems, secrets)
    finished_at = datetime.now(UTC)
    # The report and the ``job_run`` row carry the *redacted* error, not the
    # fetcher's own. ``PollResult.error`` is whatever Home Assistant or httpx
    # said, and both can quote a URL somebody put credentials in.
    detail = {**result.to_dict(), "error": error, "ok": result.ok and not error}
    if connection is not None:
        kind, name = HA_INTEGRATION
        if result.ok:
            await store.record_ok(connection, kind, name, finished_at)
        if error:
            await store.record_error(connection, kind, name, error)
        await store.record_job_run(
            connection,
            "poll_home_assistant",
            started_at=started_at,
            finished_at=finished_at,
            ok=result.ok and not error,
            detail=detail,
        )

    return _report(
        started_at,
        ok=result.ok and not error,
        error=error,
        sources=len(sources),
        polled=len(due),
        stored=stored,
        problems=problems,
        finished_at=finished_at,
        # ``ok`` and ``error`` are decided above, over the whole run rather
        # than over the fetch alone: a hub that answered and produced nothing
        # is a failure the fetch does not know about.
        **{k: v for k, v in detail.items() if k not in {"ok", "error"}},
    )


def _error_for(
    result: PollResult,
    due: list[SensorSource],
    problems: list[str],
    secrets: tuple[str, ...],
) -> str | None:
    """What belongs on ``integration.last_error`` after this run.

    Three cases, in order of how much they tell an operator:

    1. Home Assistant could not be reached — say that, and nothing else.
    2. It answered and **not one** mapped entity produced a reading. That is
       a working hub and a broken configuration, and it is the case that would
       otherwise pass as success: the job ran, no exception was raised, and
       the hypertable stayed empty.
    3. Some entities were refused, or a row is misconfigured. Named, because a
       partly-working mapping is the one people never get round to fixing.

    The result is always redacted. This string is rendered in the Ministry
    Office, and a Home Assistant error can quote a URL somebody put
    credentials in.
    """
    if result.error:
        return redact(result.error, secrets)
    reasons: list[str] = list(problems)
    if due and not result.readings:
        detail = ", ".join(
            sorted({f"{item.entity_id}: {item.reason}" for item in result.skipped})
        )
        reasons.insert(
            0,
            "Home Assistant answered, but no mapped entity produced a reading"
            + (f" ({detail})" if detail else ""),
        )
    elif result.skipped:
        detail = ", ".join(
            sorted({f"{item.entity_id}: {item.reason}" for item in result.skipped})
        )
        reasons.append(f"some entities produced no reading ({detail})")
    return redact("; ".join(reasons), secrets) if reasons else None


async def check_home_assistant(ctx: dict[str, Any]) -> dict[str, Any]:
    """Ask Home Assistant whether it is there, and record the answer.

    Separate from the poll on purpose. A deployment with no sources configured
    yet still needs the Ministry Office to say whether the hub is reachable —
    that is the screen someone looks at *while* setting it up, and a health
    check that only runs when there is something to poll is a health check
    that is absent exactly when it is wanted.
    """
    from . import store
    from .factory import get_source

    settings = _settings(ctx)
    started_at = datetime.now(UTC)
    source_api = ctx.get("source") or get_source()
    connection = ctx.get("connection")
    kind, name = HA_INTEGRATION

    if not settings.is_configured and not settings.mock_mode:
        return {
            "job": "check_home_assistant",
            "ok": False,
            "status": "unconfigured",
            "error": "MOH_HA_BASE_URL and MOH_HA_TOKEN are not set",
        }

    try:
        message = await source_api.probe()
    except HubUnavailable as exc:
        error = redact(str(exc), _secrets(settings)) or str(exc)
        if connection is not None:
            await store.record_error(connection, kind, name, error)
        return {
            "job": "check_home_assistant",
            "ok": False,
            "status": "down",
            "error": error,
        }

    finished_at = datetime.now(UTC)
    if connection is not None:
        await store.record_ok(connection, kind, name, finished_at)
    return {
        "job": "check_home_assistant",
        "ok": True,
        "status": "ok",
        "message": message,
        "took_ms": round((finished_at - started_at).total_seconds() * 1000, 1),
        "is_mock": getattr(source_api.fetcher, "is_mock", False),
    }


async def hub_health(ctx: dict[str, Any]) -> dict[str, Any]:
    """What the Ministry Office screen shows: is the hub actually answering?

    The report distinguishes four things a single "is it up" boolean cannot:
    an integration nobody has configured, one that is configured and failing,
    one that has not been *heard from* — no error, no recent success, which is
    what a stopped worker looks like — and a source that is configured, fine
    as far as the hub is concerned, and has not produced a reading in hours.
    """
    from . import store, world

    settings = _settings(ctx)
    now = datetime.now(UTC)
    connection = ctx.get("connection")

    if connection is not None:
        rows = await store.read_integrations(connection)
        integrations = [IntegrationHealth.from_row(dict(row)) for row in rows]
    else:
        integrations = world.integrations(settings)

    sources = await _sources_for(ctx)
    return health_report(
        integrations,
        [SourceHealth.from_source(source) for source in sources],
        now=now,
        floor_s=settings.source_stale_after_s,
    )


async def publish_to_mqtt(ctx: dict[str, Any]) -> dict[str, Any]:
    """Publish the Ministry's state back into Home Assistant over MQTT.

    ``contracts/events/mqtt.md`` is the contract: retained discovery configs,
    retained state, one availability topic that everything hangs off. Every
    message goes through :func:`~workers.hub.mqtt.guard`, which refuses a
    payload carrying anything credential-shaped — the contract's "nothing
    secret goes in a payload", enforced rather than remembered.

    The live transport is not built: ``api/pyproject.toml`` carries no MQTT
    client and Workstream F does not own it (rule 2), so adding ``aiomqtt`` is
    a request to A and B and is in this sprint's pull request. With no
    publisher in ``ctx`` this job builds every message, guards it, and reports
    what it *would* send — which is what the tests assert, and which is the
    useful half anyway: the bytes on the topic, not the library that carried
    them.

    **S4 fills the seam this job was left with.** In S3 nothing put ``rounds``
    or ``frost`` into ``ctx``, so a deployed worker published a permanent zero
    on ``herbology/rounds/due`` and a permanent ``OFF`` on the frost
    binary_sensor — the contract's topics, carrying nothing, which is precisely
    the silent failure this workstream exists to prevent. It now reads the same
    two endpoints the notification jobs read
    (:mod:`workers.hub.notify.state`), so a household can hang a Home Assistant
    automation off the frost sensor and have it mean something.

    A read that fails is reported and **nothing is published**. Publishing a
    zero because the API did not answer would retain that zero on the broker
    until something replaced it, which is worse than an entity going
    unavailable: the availability topic can say "we are not answering", and a
    retained zero cannot be taken back.
    """
    rounds = ctx.get("rounds")
    frost = ctx.get("frost")
    read_error: str | None = None
    if rounds is None or frost is None:
        from .notify.state import build_reader

        settings = _settings(ctx)
        reader = ctx.get("reader") or build_reader(settings)
        try:
            if rounds is None:
                rounds = await reader.rounds()
            if frost is None:
                frost = await reader.frost()
        except Exception as exc:  # noqa: BLE001 - reported, never raised at Arq
            read_error = redact(
                f"could not read the Ministry's state: {exc}", _secrets(settings)
            )
    if read_error:
        return {
            "job": "publish_to_mqtt",
            "ok": False,
            "error": read_error,
            "published": 0,
            "topics": [],
        }

    rounds = rounds or {}
    frost = frost or {}
    specimens = ctx.get("specimens") or []
    version = str(ctx.get("sw_version") or "1.0.0")

    messages = [availability(True)]
    messages += discovery_messages(
        specimen_ids=[
            str(row.get("specimen_id") or row.get("id")) for row in specimens
        ],
        sw_version=version,
    )
    messages += rounds_messages(**_rounds_state(rounds))
    messages += frost_messages(**_frost_state(frost))
    for row in specimens:
        messages += specimen_messages(
            str(row.get("specimen_id") or row.get("id")),
            water_due=bool(row.get("water_due")),
            deficit_mm=row.get("deficit_mm"),
        )

    publisher = ctx.get("publisher") or MemoryPublisher()
    sent = await publish_all(publisher, messages)
    return {
        "job": "publish_to_mqtt",
        "ok": True,
        "published": len(sent),
        "retained": sum(1 for message in sent if message.retain),
        "topics": [message.topic for message in sent],
        "transport": type(publisher).__name__,
    }


def _rounds_state(rounds: Mapping[str, Any]) -> dict[str, Any]:
    """``MorningRounds`` into the three things the contract's topics carry.

    Accepts the already-counted shape too (``{"due": 4, "overdue": 1}``), which
    is what a caller with the numbers to hand passes and what S3's tests use.
    A ``due`` that is a list is counted; a ``due`` that is a number is taken as
    the count.

    ``overdue`` is derived from the tasks themselves rather than asked for:
    ``MorningRounds`` has no overdue field, and a task whose ``due_at`` is
    before today is what "overdue" means.
    """
    raw = rounds.get("due")
    if isinstance(raw, (int, float)) and not isinstance(raw, bool):
        return {
            "due": int(raw),
            "overdue": int(rounds.get("overdue") or 0),
            "tasks": rounds.get("tasks") or (),
        }
    rows = raw if isinstance(raw, Sequence) and not isinstance(raw, str) else ()
    tasks = [task for task in rows if isinstance(task, Mapping)]
    today = str(rounds.get("date") or "")
    overdue = sum(
        1 for task in tasks if today and str(task.get("due_at") or "")[:10] < today
    )
    return {
        "due": len(tasks),
        "overdue": overdue,
        "tasks": [_task_row(task) for task in tasks],
    }


def _task_row(task: Mapping[str, Any]) -> dict[str, Any]:
    """The four fields ``mqtt._task_attributes`` allows, named as it expects.

    The allow-list there is the guard; this is the translation. Rule 7's plain
    title is what a dashboard shows, for the same reason a notification does.
    """
    specimen = task.get("specimen")
    name = ""
    if isinstance(specimen, Mapping):
        name = str(specimen.get("nickname") or specimen.get("display_name") or "")
    return {
        "id": task.get("id", ""),
        "specimen": name,
        "kind": task.get("plain_title") or task.get("task_type") or "",
        "due_on": str(task.get("due_at") or "")[:10] or None,
    }


def _frost_state(frost: Mapping[str, Any]) -> dict[str, Any]:
    """``FrostReport`` into the frost binary_sensor and the next-frost date.

    ``unassessable`` deliberately does not turn the sensor ``ON``. A plant that
    could not be judged is not a plant known to be at risk, and an automation
    that closed the greenhouse vents on "we do not know" would be acting on
    nothing. It is carried in the attributes instead, where a dashboard can
    show it and ``herbology/frost/next`` stays ``unknown`` rather than
    becoming a date nobody computed.
    """
    if "active" in frost:
        return {
            "active": bool(frost.get("active")),
            "night": frost.get("night"),
            "low_c": frost.get("low_c"),
            "specimens": frost.get("specimens") or (),
            "next_frost": frost.get("next"),
        }
    alerts = [
        alert
        for alert in (frost.get("alerts") or ())
        if isinstance(alert, Mapping) and str(alert.get("state") or "open") == "open"
    ]
    nights = sorted(
        str(alert.get("night_of")) for alert in alerts if alert.get("night_of")
    )
    lows = [
        float(alert["forecast_low_c"])
        for alert in alerts
        if isinstance(alert.get("forecast_low_c"), (int, float))
    ]
    names = []
    for alert in alerts:
        specimen = alert.get("specimen")
        if isinstance(specimen, Mapping):
            names.append(
                str(specimen.get("nickname") or specimen.get("display_name") or "")
            )
    return {
        "active": bool(alerts),
        "night": nights[0] if nights else None,
        "low_c": min(lows) if lows else None,
        "specimens": [name for name in names if name],
        "next_frost": nights[0] if nights else None,
    }


def _report(
    started_at: datetime,
    *,
    ok: bool,
    error: str | None = None,
    finished_at: datetime | None = None,
    **extra: Any,
) -> dict[str, Any]:
    finished = finished_at or datetime.now(UTC)
    return {
        "job": "poll_home_assistant",
        "ok": ok,
        "error": error,
        "started_at": started_at.isoformat(),
        "took_ms": round((finished - started_at).total_seconds() * 1000, 1),
        **extra,
    }


class WorkerSettings(metaclass=ArqBootstrap):
    """Arq entrypoint.

    The schedules follow the sprint's exit criterion — indoor readings stored
    every 5–15 minutes. Five for the poll, which is the finest cadence any row
    asks for; five for the health check, offset by two minutes so the Ministry
    Office learns the hub is down from a cheap ``GET /api/`` rather than from
    a poll that had to time out first.
    """

    functions: ClassVar[list] = [
        poll_home_assistant,
        check_home_assistant,
        hub_health,
        publish_to_mqtt,
        notify_rounds,
        notify_frost,
        notify_integration_health,
    ]

    cron_jobs: ClassVar[list] = []


def _build_cron_jobs() -> list[Any]:
    """Built lazily so importing this module never requires Arq to be installed.

    :data:`SUPPORTED_METRICS` and the topic helpers are imported by other
    workstreams and by the mocks, neither of which runs jobs. A missing
    scheduler dependency must not take the adapter's constants down with it.
    """
    try:
        from arq import cron
    except ImportError:  # pragma: no cover - arq is a declared dependency
        return []

    return [
        cron(poll_home_assistant, minute=set(range(0, 60, 5)), run_at_startup=True),
        cron(
            check_home_assistant,
            minute={2, 7, 12, 17, 22, 27, 32, 37, 42, 47, 52, 57},
            run_at_startup=True,
        ),
        # After the poll and after G's rules have run for the morning.
        cron(publish_to_mqtt, minute={10, 40}),
        # --- notifications (S4) ------------------------------------------
        #
        # Frost every fifteen minutes, and that cadence is the requirement
        # rather than a preference: the brief's own example is a Freeze
        # Warning issued at 2 PM for that night, and a daily job learns about
        # it the following morning. The job is cheap — one GET, and nothing
        # sent unless the forecast is new news — so the interval is set by how
        # late an alert may be, not by how much it costs.
        cron(notify_frost, minute={0, 15, 30, 45}),
        # Rounds hourly: the job itself decides whether the site's local clock
        # has reached the hour the member asked for, because a cron in UTC
        # cannot know that and a household that moves through a DST change
        # would otherwise get their rounds an hour out twice a year.
        cron(notify_rounds, minute={5}),
        # Health hourly, well after the poll and the health check have had a
        # chance to run and record. It waits out its own grace period on top.
        cron(notify_integration_health, minute={25}),
    ]


WorkerSettings.cron_jobs = _build_cron_jobs()

File: web/src/routes/rounds/tasks.ts
/** The arithmetic and the wording of Morning Rounds, as pure functions.
 *
 * The screen itself is markup; everything that decides what a reader is told
 * lives here, where it can be tested without a browser. Three rules shape most
 * of what follows, and all three are about not letting the screen claim more
 * than the API said:
 *
 *   1. **Satisfied is not done.** A watering the rain settled stays on the
 *      round as settled. It is not a completed job and it is not an absent one.
 *   2. **A silence is rendered.** A plant the scheduler could not schedule, and
 *      an API that does not say whether there were any, both get a sentence.
 *   3. **A guess never arrives looking like a measurement.** Under ADR 0010
 *      nothing in this house measures the soil, so a task built on an uncited
 *      interval or a degraded water balance says so next to the instruction.
 */

import { STATUS, type Status, type SelectOption } from '$ui';
import { reportsItsOwnConfidence } from '../shared/assessment';
import { taskTypeLabel } from '../shared/labels';
import type {
  Confidence,
  Member,
  MorningRounds,
  SpecimenName,
  Task,
  UnscheduledSpecimen,
} from './api';

/** Whole days between two instants, floored — "overdue by 1 day" needs a full
 *  day to have passed, not a minute past midnight. */
export function daysBetween(from: string, to: Date): number {
  const due = new Date(from);
  if (Number.isNaN(due.getTime())) return 0;
  const startOfDue = Date.UTC(due.getUTCFullYear(), due.getUTCMonth(), due.getUTCDate());
  const startOfNow = Date.UTC(to.getUTCFullYear(), to.getUTCMonth(), to.getUTCDate());
  return Math.round((startOfNow - startOfDue) / 86_400_000);
}

/** When this was owed, in plain words. Morning Rounds shows what is due today
 *  and everything still owed from before, and those two are not the same thing
 *  to a reader deciding what to carry outside. */
export function dueSentence(task: Task, now: Date): string {
  const late = daysBetween(task.due_at, now);
  if (late <= 0) return 'due today';
  if (late === 1) return 'a day overdue';
  return `${late} days overdue`;
}

/** The plain second line of a task row: the job, when it was owed, how much. */
export function taskMeta(task: Task, now: Date): string {
  const parts = [taskTypeLabel(task.task_type), dueSentence(task, now)];
  if (task.amount_ml) parts.push(`${task.amount_ml} ml`);
  return parts.join(' · ');
}

/** What settled a watering, as a pill. Three of the four have a `$ui` status
 *  already; `forecast_change` and `manual` are this screen's, paired the same
 *  way. Never "done": the job was not done, it stopped being needed. */
export function satisfiedStatus(task: Task): Status {
  switch (task.satisfied_by) {
    case 'rain':
      return STATUS.satedByRain;
    case 'sensor':
      return STATUS.satedBySensor;
    case 'forecast_change':
      return {
        themed: 'The heavens changed their mind',
        plain: 'The forecast changed — no longer needed',
        tone: 'sated',
      };
    case 'manual':
      return {
        themed: 'Set aside by hand',
        plain: 'Marked as not needed by a person',
        tone: 'sated',
      };
    default:
      return {
        themed: 'Settled, cause unrecorded',
        plain: 'Settled — the API did not say what settled it',
        tone: 'sated',
      };
  }
}

/**
 * How much a task's instruction is worth, as a pill.
 *
 * Deliberately not the Almanac's vocabulary. "Read through cloud" is the right
 * phrase for a forecast whose ingest is a day stale; it is the wrong phrase for
 * a watering scheduled off an interval nobody ever cited, which is not a
 * reading at all. Same four levels (ADR 0004), different sentence — the same
 * split the Specimen page already makes between a measured value and an
 * attested one.
 *
 * `null` is its own answer and is not `high`: a completed task cannot say how
 * sure it was, because the frozen `task` table has no column its confidence
 * could have been kept in (G's escalation 2).
 */
export function taskConfidence(confidence: Confidence | null | undefined): Status {
  switch (confidence) {
    case 'high':
      return { themed: 'Set down with certainty', plain: 'High confidence', tone: 'thriving' };
    case 'medium':
      return { themed: 'Set down with care', plain: 'Medium confidence', tone: 'sated' };
    case 'low':
      return { themed: 'A rough reckoning', plain: 'Low confidence — check it', tone: 'parched' };
    case 'unknown':
      return {
        themed: 'A guess, and said to be one',
        plain: 'Confidence unknown — nothing attests this',
        tone: 'ailing',
      };
    default:
      return {
        themed: 'Not stated',
        plain: 'This task does not say how sure it is',
        tone: 'parched',
      };
  }
}

/** Why a settled task is still on the screen. Said once per section, not once
 *  per row: it is an explanation of the section's existence. */
export const SATISFIED_EXPLANATION =
  'Nothing to do here. These waterings were owed today and something else covered them — ' +
  'they stay on the round so the day reads as a complete record rather than as a short one.';

/** Does this payload say anything at all about plants it could not schedule?
 *
 *  `unscheduled: []` means "every plant is accounted for". A missing field
 *  means the API never said, and an API that never said is exactly the silence
 *  the list exists to break — so the screen says so instead of showing nothing,
 *  the same way the Almanac does for a forecast that states no confidence. */
export function reportsUnscheduled(rounds: MorningRounds | null | undefined): boolean {
  return Array.isArray(rounds?.unscheduled);
}

export const UNSCHEDULED_SILENCE =
  'This build of the API does not report the plants it could not schedule, so this screen cannot ' +
  'tell you whether any were left off. A plant with no watering interval generates no task, and a ' +
  'plant with no task looks exactly like a plant that needs nothing. Added by Workstream G and ' +
  'asked of Workstream A; until the contract carries it, read the rounds below as "what was ' +
  'scheduled", not as "every plant in the house".';

/** Do these tasks carry their own certainty?
 *
 *  Asked of the whole list rather than of each row: one notice at the top of a
 *  section is a warning, and nine of them is wallpaper. A single task that
 *  reports is enough to show the API knows the fields. */
export function tasksReportCertainty(tasks: readonly Task[]): boolean {
  return tasks.some((task) => reportsItsOwnConfidence(task));
}

export const CERTAINTY_SILENCE =
  'These tasks do not say how sure the scheduler is about them. A watering worked out from an ' +
  'uncited interval and one measured for the species look identical here. Workstream G serves ' +
  'that certainty today and the contract does not carry it yet — asked of Workstream A.';

/** The caveat to print under an instruction when there is no structured one.
 *
 *  `detail` is in the frozen contract and `degradations[]` is not, so G writes
 *  the same caveat into both: the plain sentence on `detail` for a client that
 *  reads neither of the new fields, and the structured reasons for one that
 *  does. Printing both would show the reader one caveat twice. The structured
 *  version wins where it exists, because it carries the ceiling each reason
 *  puts on the answer; `detail` is what is left when it does not — which is the
 *  case for every completed task, whose confidence the frozen `task` table has
 *  no column to have kept.
 */
export function instructionNote(task: Task): string | null {
  if (task.degradations?.length) return null;
  return task.detail?.trim() || null;
}

/** A plant left off the rounds, given a name where the register has one.
 *
 *  `unscheduled[]` carries a bare `specimen_id`, so a name costs a second read
 *  of the register. When that read fails the reason is still shown — a uuid
 *  with a reason beside it is worth more than a plant quietly missing. */
export interface UnscheduledRow {
  specimenId: string;
  name: string;
  /** False when the name is a stand-in, so the row can say so plainly. */
  named: boolean;
  reason: string;
  href: string;
}

export function unscheduledRows(
  unscheduled: readonly UnscheduledSpecimen[],
  specimens: readonly SpecimenName[] | null,
): UnscheduledRow[] {
  const names = new Map((specimens ?? []).map((s) => [s.id, s.display_name]));
  return unscheduled.map((entry) => {
    const name = names.get(entry.specimen_id);
    return {
      specimenId: entry.specimen_id,
      name: name ?? 'A plant the register could not name just now',
      named: Boolean(name),
      reason: entry.reason,
      href: `/specimen/${entry.specimen_id}/tending`,
    };
  });
}

// --------------------------------------------------------------- completion

/** How the household member is remembered between visits.
 *
 *  ADR 0008: the picker defaults to the last member used **on that device**,
 *  which is a per-device preference and not a server-side identity. So it lives
 *  in `localStorage` and nowhere else. */
export const MEMBER_STORAGE_KEY = 'moh:last-member';

export function readRememberedMember(): string | null {
  try {
    return globalThis.localStorage?.getItem(MEMBER_STORAGE_KEY) ?? null;
  } catch {
    // Private browsing, or storage switched off. A forgotten default is a
    // nuisance; a screen that will not load because of one is a defect.
    return null;
  }
}

export function rememberMember(id: string): void {
  try {
    globalThis.localStorage?.setItem(MEMBER_STORAGE_KEY, id);
  } catch {
    /* see above */
  }
}

/** Who the picker starts on: the remembered member if they are still in the
 *  household, otherwise the first one, otherwise nobody. */
export function initialMember(members: readonly Member[], remembered: string | null): string {
  if (remembered && members.some((member) => member.id === remembered)) return remembered;
  return members[0]?.id ?? '';
}

export function memberOptions(members: readonly Member[]): SelectOption[] {
  return members.map((member) => ({ value: member.id, plain: member.name }));
}

export function memberName(members: readonly Member[], id: string): string | null {
  return members.find((member) => member.id === id)?.name ?? null;
}

/** What the live region announces after a batch is marked done.
 *
 *  Plain language only — this is read aloud, and a themed phrase read aloud on
 *  its own is the pairing rule broken where it matters most. The member is
 *  named because attribution is the point of asking (ADR 0008). */
export function completionAnnouncement(count: number, who: string | null): string {
  const jobs = count === 1 ? '1 task' : `${count} tasks`;
  return who
    ? `${jobs} marked done, recorded against ${who}.`
    : `${jobs} marked done. No household member was recorded.`;
}

/** What the batch button offers to do, in plain words, so the label says how
 *  many rather than leaving the reader to count their own ticks. */
export function batchButtonPlain(count: number): string {
  if (count === 0) return 'Mark done — nothing selected yet';
  return count === 1 ? 'Mark 1 task done' : `Mark ${count} tasks done`;
}

/** The sentence shown when a batch did not go through. The selection is kept,
 *  so the sentence says so: a reader who has just ticked six rows should not
 *  have to find them again. */
export function completionFailure(problem: string): string {
  return `Nothing was marked done. ${problem} Your selection has been kept — try again.`;
}

/** Ids worth sending: everything selected that is still on the due list.
 *
 *  A tick left over from a task that has since been settled by rain, or
 *  completed on somebody else's phone, is dropped rather than sent. G skips
 *  unknown ids instead of refusing the batch, which is the right server
 *  behaviour and not a reason for the client to send nonsense. */
export function completableIds(due: readonly Task[], selected: ReadonlySet<string>): string[] {
  return due.filter((task) => selected.has(task.id)).map((task) => task.id);
}
